sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,553 @@
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct sub_group_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<sub_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct work_item_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_item_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_group_progress_key::value_t<Guarantee, CoordinationScope> work_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr sub_group_progress_key::value_t<Guarantee, CoordinationScope> sub_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_item_progress_key::value_t<Guarantee, CoordinationScope> work_item_progress;
+
+template <> struct is_property_key<work_group_progress_key> : std::true_type {};
+template <> struct is_property_key<sub_group_progress_key> : std::true_type {};
+template <> struct is_property_key<work_item_progress_key> : std::true_type {};
+
+}
+----
+
+|===
+|Property|Description
+
+|`work_group_progress`
+|The `work_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_group` scope, with the specified
+ coordination scope.
+
+|`sub_group_progress`
+|The `sub_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::sub_group` scope, with the specified coordination
+ scope.
+
+|`work_item_progress`
+|The `work_item_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_item` scope, with the specified coordination
+ scope.
+
+|===
+
+
+=== Launch queries [[launch]]
+
+An implementation's ability to satisify requests for specific forward progress
+guarantees for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
```

Comment:
Perhaps this has been considered already but should we also have queries that are the reverse, which allow you to ask a device or a kernel if they can provide a particular forward progress guarantee for a particular coordination scope and target scope?

Reply:
The intent was that this would be covered by the device query.

Rather than getting a Boolean value for each (forward progress guarantee, coordination scope, target scope) triple, the query returns a forward progress guarantee for each (coordination scope, target scope) pair.

The reasoning behind that decision is that if an implementation can a provide strong forward progress guarantees at some scope, it can trivially provide weaker forward progress guarantees at the same scope.  For example, because a thread providing concurrent forward progress guarantees also provides parallel and weakly parallel forward progress guarantees, an implementation could just use its concurrent implementation for all kernels.

Reply:
As noted below, I've since updated this so that the device queries return the full set of forward progress guarantees supported for a particular coordination scope and target scope.

Reply:
That's great, this covers the use case I had in mind, thanks!